dask/array/random.py

```diff
@@ -17,7 +17,603 @@
-from dask.utils import derived_from, random_state_data
+from dask.utils import derived_from, random_rng_data, random_state_data
+
+
+class Generator:
+    """
+    Container for the BitGenerators.
+
+    ``Generator`` exposes a number of methods for generating random
+    numbers drawn from a variety of probability distributions. In addition to
+    the distribution-specific arguments, each method takes a keyword argument
+    `size` that defaults to ``None``. If `size` is ``None``, then a single
+    value is generated and returned. If `size` is an integer, then a 1-D
+    array filled with generated values is returned. If `size` is a tuple,
+    then an array with that shape is filled and returned.
+
+    The function :func:`dask.array.random.default_rng` will instantiate
+    a `Generator` with numpy's default `BitGenerator`.
+
+    **No Compatibility Guarantee**
+
+    ``Generator`` does not provide a version compatibility guarantee. In
+    particular, as better algorithms evolve the bit stream may change.
+
+    This object is identical to ``np.random.Generator`` except that all
+    functions also take a ``chunks=`` keyword argument.
+
+    Parameters
+    ----------
+    bit_generator : BitGenerator
+        BitGenerator to use as the core generator.
+
+    Notes
+    -----
+    The Python stdlib module `random` contains pseudo-random number generator
+    with a number of methods that are similar to the ones available in
+    ``Generator``. It uses Mersenne Twister, and this bit generator can
+    be accessed using ``MT19937``. ``Generator``, besides being
+    Dask-aware, has the advantage that it provides a much larger number
+    of probability distributions to choose from.
+
+    Examples
+    --------
+    >>> from numpy.random import PCG64
+    >>> from dask.array.random import Generator
+    >>> rng = Generator(PCG64())
+    >>> rng.standard_normal().compute() # doctest: +SKIP
+    array(0.44595957)  # random
+
+    See Also
+    --------
+    default_rng : Recommended constructor for `Generator`.
+    np.random.Generator
+    """
+
+    def __init__(self, bit_generator):
+        self._bit_generator = bit_generator
+        self._bit_generator_state = bit_generator.state
+        self._generator = Generator
+
+    def __str__(self):
+        _str = self.__class__.__name__
+        _str += "(" + self._bit_generator.__class__.__name__ + ")"
+        return _str
+
+    def _wrap(
+        self, funcname, *args, size=None, chunks="auto", extra_chunks=(), **kwargs
+    ):
+        """Wrap numpy random function to produce dask.array random function
+
+        extra_chunks should be a chunks tuple to append to the end of chunks
+        """
+        if size is not None and not isinstance(size, (tuple, list)):
+            size = (size,)
+
+        shapes = list(
+            {
+                ar.shape
+                for ar in chain(args, kwargs.values())
+                if isinstance(ar, (Array, np.ndarray))
```

Comment:
What happens if it is not one of these types (like if it is a nested `list` or similar)? Don't we still want to determine its `shape` for broadcasting?

Also seems like we are going to a lot of work below to handle different array types. What about just coercing everything to Dask Arrays and moving on?

Reply:
Comment for a few of you comments:

That is code from `da.random.RandomState`. I tried to keep those bits as they are and not change them because a/ I am not familiar enough with the code to understand any (or if there is any) side effects b/ I tried to keep the changeset as small as possible to make it easier to review and merge

Once (if) this PR is merged, next steps would be 1/ change internal `RandomState` usage to `Generator` 2/ get feedback and work on the `Generator` code 3/ deprecate `RandomState` with suitable time intervals between steps.

But until then, I prefer to keep these bits as they are. I feel that we are already making a lot changes

Reply:
Is there a way to reuse that code directly without copying it?

Reply:
done